File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/H._Глобальная_замена/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    std::string s, p, t;
    std::cin >> s >> p >> t;
    int n = p.size();
    std::vector<int> matching;

    std::string S = p + '#' + s;
    int prevK = 0;
    int m = S.size();
    std::vector<int> lcp (p.size(), 0);
    for (int i = 1; i < m; ++i) {
        int k = prevK;
        while (k && (S[k] != S[i])) k = lcp[k - 1];
        if (S[k] == S[i]) ++k;
        if (i < n) lcp[i] = k;
        prevK = k;
        if (k == n) matching.push_back(i - 2 * n);
    }

    std::reverse(matching.begin(), matching.end());
    int i = 0;
    while (i < s.size()) {
        if (!matching.empty() && i == matching.back()) {
            std::cout << t;
            matching.pop_back();
            i += p.size();
        } else {
            std::cout << s[i];
            ++i;
        }
    }
    std::cout << "\n";

}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/H._Глобальная_замена/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    s: str = sys.stdin.readline().rstrip()
    p: str = sys.stdin.readline().rstrip()
    t: str = sys.stdin.readline().rstrip()
    sentinel: str = ''.join([p, '#', s])
    matching: list[int] = []
    lcp: list[int] = [0] * len(p)
    prev_k: int = 0
    for i in range(1, len(sentinel)):
        k: int = prev_k
        while k and sentinel[k] != sentinel[i]: k = lcp[k - 1]
        if sentinel[k] == sentinel[i]: k += 1
        if i < len(p): lcp[i] = k
        prev_k = k
        if k == len(p): matching.append(i - 2 * len(p))
    matching.reverse()
    i: int = 0
    while i < len(s):
        if matching and matching[-1] == i:
            sys.stdout.write(t)
            matching.pop()
            i += len(p)
        else:
            sys.stdout.write(s[i])
            i += 1
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution()